## ECBM E6040 Homework 4 - Programming Problem

In [1]:
import matplotlib
import pylab
from hw4a import *
from hw4b import *
%matplotlib inline

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

Testing accuracy was best for the model with 200 hidden units, 5 words in the context window, and embedding dimension 100. The worst model had parameters nhidden=100, win=7, emb_dimension=100. Most of the results were similar though for all the models I ran, ranging between a test F1 greater than 93.4 and less than 94.1. I ran all permutations of embedding dimension in (50,100), nhidden in (200,100), and win in (5,7). I ran the experiments for only 25 epochs because it ran very slowly. 

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.18 best test F1 66.22
NEW BEST: epoch 1 valid F1 80.63 best test F1 77.98
NEW BEST: epoch 2 valid F1 83.79 best test F1 80.66
NEW BEST: epoch 3 valid F1 88.42 best test F1 84.92
NEW BEST: epoch 4 valid F1 89.65 best test F1 85.97
NEW BEST: epoch 5 valid F1 91.02 best test F1 86.87
NEW BEST: epoch 6 valid F1 91.45 best test F1 88.81
NEW BEST: epoch 7 valid F1 92.05 best test F1 89.89
NEW BEST: epoch 8 valid F1 93.78 best test F1 90.69
NEW BEST: epoch 16 valid F1 95.75 best test F1 93.31
NEW BEST: epoch 17 valid F1 96.16 best test F1 93.35
NEW BEST: epoch 18 valid F1 96.36 best test F1 93.3

NEW BEST: epoch 20 valid F1 96.37 best test F1 93.39
NEW BEST: epoch 21 valid F1 96.51 best test F1 93.97
NEW BEST: epoch 22 valid F1 96.56 best test F

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 71.53 best test F1 67.86
NEW BEST: epoch 1 valid F1 81.24 best test F1 78.63
NEW BEST: epoch 2 valid F1 84.62 best test F1 81.91
NEW BEST: epoch 3 valid F1 88.65 best test F1 84.7
NEW BEST: epoch 4 valid F1 89.71 best test F1 86.27
NEW BEST: epoch 5 valid F1 91.82 best test F1 87.68

NEW BEST: epoch 7 valid F1 93.15 best test F1 90.95
NEW BEST: epoch 8 valid F1 94.2 best test F1 91.9
NEW BEST: epoch 9 valid F1 94.44 best test F1 91.73
NEW BEST: epoch 10 valid F1 95.07 best test F1 92.32

NEW BEST: epoch 12 valid F1 95.55 best test F1 93.1
NEW BEST: epoch 13 valid F1 95.64 best test F1 93.01


NEW BEST: epoch 16 valid F1 95.65 best test F1 93.37
NEW BEST: epoch 17 valid F1 96.47 best test F1 93.79




NEW BEST: epoch 22 valid F1 96.68 best t

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.47 best test F1 65.79
NEW BEST: epoch 1 valid F1 80.18 best test F1 76.07
NEW BEST: epoch 2 valid F1 82.58 best test F1 79.6
NEW BEST: epoch 3 valid F1 87.23 best test F1 83.35
NEW BEST: epoch 4 valid F1 89.3 best test F1 85.58
NEW BEST: epoch 5 valid F1 90.55 best test F1 87.2

NEW BEST: epoch 7 valid F1 91.56 best test F1 89.85
NEW BEST: epoch 8 valid F1 93.18 best test F1 91.09
NEW BEST: epoch 9 valid F1 93.7 best test F1 91.58
NEW BEST: epoch 10 valid F1 94.3 best test F1 91.64

NEW BEST: epoch 12 valid F1 94.86 best test F1 92.29
NEW BEST: epoch 13 valid F1 94.87 best test F1 92.42
NEW BEST: epoch 14 valid F1 95.4 best test F1 92.36


NEW BEST: epoch 17 valid F1 95.91 best test F1 93.41
NEW BEST: epoch 18 valid F1 95.96 best test F1 

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.76 best test F1 68.63
NEW BEST: epoch 1 valid F1 80.95 best test F1 78.45
NEW BEST: epoch 2 valid F1 84.54 best test F1 81.6
NEW BEST: epoch 3 valid F1 87.75 best test F1 84.64
NEW BEST: epoch 4 valid F1 89.6 best test F1 86.41
NEW BEST: epoch 5 valid F1 91.2 best test F1 87.78

NEW BEST: epoch 7 valid F1 92.58 best test F1 90.04
NEW BEST: epoch 8 valid F1 93.78 best test F1 91.27
NEW BEST: epoch 9 valid F1 94.37 best test F1 91.22
NEW BEST: epoch 10 valid F1 95.04 best test F1 92.06
NEW BEST: epoch 11 valid F1 95.37 best test F1 92.75
NEW BEST: epoch 12 valid F1 95.58 best test F1 92.72


NEW BEST: epoch 15 valid F1 96.35 best test F1 93.51

NEW BEST: epoch 17 valid F1 96.63 best test F1 93.45



NEW BEST: epoch 21 valid F1 96.94 best t

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.35 best test F1 60.56
NEW BEST: epoch 1 valid F1 75.35 best test F1 72.82
NEW BEST: epoch 2 valid F1 80.31 best test F1 77.08
NEW BEST: epoch 3 valid F1 83.77 best test F1 80.65
NEW BEST: epoch 4 valid F1 85.97 best test F1 82.57
NEW BEST: epoch 5 valid F1 88.01 best test F1 84.32
NEW BEST: epoch 6 valid F1 89.31 best test F1 85.13
NEW BEST: epoch 7 valid F1 91.03 best test F1 88.09
NEW BEST: epoch 8 valid F1 91.69 best test F1 89.0
NEW BEST: epoch 9 valid F1 92.58 best test F1 89.71
NEW BEST: epoch 10 valid F1 93.13 best test F1 90.29
NEW BEST: epoch 11 valid F1 93.89 best test F1 91.14
NEW BEST: epoch 12 valid F1 94.12 best test F1 91.57
NEW BEST: epoch 13 valid F1 94.58 best test F1 92.08
NEW BEST: epoch 14 valid F1 94.74 best test F1 

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 65.82 best test F1 64.7
NEW BEST: epoch 1 valid F1 78.09 best test F1 74.95
NEW BEST: epoch 2 valid F1 82.15 best test F1 78.23
NEW BEST: epoch 3 valid F1 85.42 best test F1 81.91
NEW BEST: epoch 4 valid F1 87.29 best test F1 85.29
NEW BEST: epoch 5 valid F1 89.18 best test F1 85.98
NEW BEST: epoch 6 valid F1 90.44 best test F1 86.88
NEW BEST: epoch 7 valid F1 91.64 best test F1 88.6
NEW BEST: epoch 8 valid F1 92.43 best test F1 89.39
NEW BEST: epoch 9 valid F1 93.27 best test F1 89.4
NEW BEST: epoch 10 valid F1 93.92 best test F1 90.85
NEW BEST: epoch 11 valid F1 94.33 best test F1 90.84
NEW BEST: epoch 12 valid F1 94.53 best test F1 92.17
NEW BEST: epoch 13 valid F1 95.15 best test F1 91.87



NEW BEST: epoch 17 valid F1 95.93 best test F

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.75 best test F1 60.45
NEW BEST: epoch 1 valid F1 74.27 best test F1 71.69
NEW BEST: epoch 2 valid F1 80.15 best test F1 77.48
NEW BEST: epoch 3 valid F1 84.54 best test F1 81.34
NEW BEST: epoch 4 valid F1 86.84 best test F1 83.14
NEW BEST: epoch 5 valid F1 88.49 best test F1 85.0
NEW BEST: epoch 6 valid F1 89.42 best test F1 85.82
NEW BEST: epoch 7 valid F1 90.62 best test F1 87.43
NEW BEST: epoch 8 valid F1 91.65 best test F1 88.42
NEW BEST: epoch 9 valid F1 92.73 best test F1 89.08
NEW BEST: epoch 10 valid F1 92.91 best test F1 90.04
NEW BEST: epoch 11 valid F1 93.41 best test F1 90.94

NEW BEST: epoch 13 valid F1 94.55 best test F1 91.35
NEW BEST: epoch 14 valid F1 94.66 best test F1 92.01
NEW BEST: epoch 15 valid F1 95.25 best test F1

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 65.5 best test F1 64.61
NEW BEST: epoch 1 valid F1 78.64 best test F1 75.28
NEW BEST: epoch 2 valid F1 82.52 best test F1 78.33
NEW BEST: epoch 3 valid F1 84.9 best test F1 82.08
NEW BEST: epoch 4 valid F1 87.5 best test F1 84.57
NEW BEST: epoch 5 valid F1 89.29 best test F1 85.25
NEW BEST: epoch 6 valid F1 90.17 best test F1 86.33
NEW BEST: epoch 7 valid F1 91.24 best test F1 89.05
NEW BEST: epoch 8 valid F1 92.13 best test F1 89.72
NEW BEST: epoch 9 valid F1 93.42 best test F1 90.3
NEW BEST: epoch 10 valid F1 93.97 best test F1 90.91
NEW BEST: epoch 11 valid F1 94.43 best test F1 91.98
NEW BEST: epoch 12 valid F1 94.74 best test F1 91.77
NEW BEST: epoch 13 valid F1 94.9 best test F1 91.9

NEW BEST: epoch 15 valid F1 95.23 best test F1 92.

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

The model actually performs better without normalization in some cases and worse in others, although only minimally (less than 1% difference), which I found surprising. Again, the best and worst model had similar results, ranging from 93.35 to 94.42 F1. Here the worst results were lower than the normalized models and the best results were higher than the best normalized model. The best model without normalization used nhidden=100, win=7, emb_dimension=100 and had 94.42 test F1. The same parameters for the best model here were the same parameters for the worst model that was normalized. The worst model without normalization had parameters nhidden=100, win=7, emb_dimension=50. I used the same combination of parameters here as in bullet 1.

In [2]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.96 best test F1 63.35
NEW BEST: epoch 1 valid F1 81.28 best test F1 76.9
NEW BEST: epoch 2 valid F1 84.73 best test F1 81.46
NEW BEST: epoch 3 valid F1 88.04 best test F1 84.06
NEW BEST: epoch 4 valid F1 90.05 best test F1 86.02
NEW BEST: epoch 5 valid F1 91.31 best test F1 86.93
NEW BEST: epoch 6 valid F1 92.65 best test F1 88.75
NEW BEST: epoch 7 valid F1 93.21 best test F1 89.79
NEW BEST: epoch 8 valid F1 94.0 best test F1 90.6
NEW BEST: epoch 9 valid F1 94.86 best test F1 90.68
NEW BEST: epoch 10 valid F1 95.13 best test F1 91.98

NEW BEST: epoch 12 valid F1 95.68 best test F1 92.92
NEW BEST: epoch 13 valid F1 95.91 best test F1 92.79

NEW BEST: epoch 15 valid F1 96.25 best test F1 93.48
NEW BEST: epoch 16 valid F1 96.33 best test F1

In [ ]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 74.89 best test F1 68.6
NEW BEST: epoch 1 valid F1 83.21 best test F1 79.66
NEW BEST: epoch 7 valid F1 94.23 best test F1 91.71
NEW BEST: epoch 8 valid F1 94.77 best test F1 91.98
NEW BEST: epoch 9 valid F1 95.27 best test F1 92.61
NEW BEST: epoch 10 valid F1 95.59 best test F1 92.57
NEW BEST: epoch 11 valid F1 95.99 best test F1 93.26
NEW BEST: epoch 12 valid F1 96.3 best test F1 93.71


NEW BEST: epoch 15 valid F1 96.65 best test F1 93.65

NEW BEST: epoch 17 valid F1 96.69 best test F1 93.82


NEW BEST: epoch 20 valid F1 96.84 best test F1 94.16

NEW BEST: epoch 22 valid F1 97.16 best test F1 93.67
NEW BEST: epoch 23 valid F1 97.29 best test F1 94.04

BEST RESULT: epoch 23 valid F1 97.29 best test F1 94.04 with the model ../result


In [6]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.35 best test F1 64.78
NEW BEST: epoch 1 valid F1 80.88 best test F1 76.34
NEW BEST: epoch 2 valid F1 84.37 best test F1 81.26
NEW BEST: epoch 3 valid F1 88.0 best test F1 84.54
NEW BEST: epoch 4 valid F1 90.16 best test F1 85.94
NEW BEST: epoch 5 valid F1 90.97 best test F1 87.24
NEW BEST: epoch 6 valid F1 91.91 best test F1 88.86
NEW BEST: epoch 7 valid F1 93.17 best test F1 90.68
NEW BEST: epoch 8 valid F1 94.13 best test F1 91.27
NEW BEST: epoch 9 valid F1 95.02 best test F1 91.69
NEW BEST: epoch 10 valid F1 95.25 best test F1 92.44
NEW BEST: epoch 11 valid F1 95.36 best test F1 92.36
NEW BEST: epoch 12 valid F1 95.74 best test F1 93.1
NEW BEST: epoch 13 valid F1 96.03 best test F1 93.01
NEW BEST: epoch 14 valid F1 96.07 best test F1 

In [5]:
emb, idx2label, idx2word = test_rnnslu(nhidden=200, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 74.25 best test F1 69.77
NEW BEST: epoch 1 valid F1 82.79 best test F1 78.64
NEW BEST: epoch 2 valid F1 86.6 best test F1 83.45
NEW BEST: epoch 3 valid F1 89.3 best test F1 85.68
NEW BEST: epoch 4 valid F1 91.0 best test F1 87.59
NEW BEST: epoch 5 valid F1 92.07 best test F1 88.85
NEW BEST: epoch 6 valid F1 92.64 best test F1 90.06
NEW BEST: epoch 7 valid F1 94.16 best test F1 91.63
NEW BEST: epoch 8 valid F1 94.52 best test F1 92.04
NEW BEST: epoch 9 valid F1 95.05 best test F1 92.0
NEW BEST: epoch 10 valid F1 95.56 best test F1 93.22
NEW BEST: epoch 11 valid F1 95.63 best test F1 93.03
NEW BEST: epoch 12 valid F1 96.07 best test F1 93.09


NEW BEST: epoch 15 valid F1 96.72 best test F1 93.41

NEW BEST: epoch 17 valid F1 96.96 best test F

In [7]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.95 best test F1 60.69
NEW BEST: epoch 1 valid F1 76.68 best test F1 72.2
NEW BEST: epoch 2 valid F1 81.7 best test F1 78.4
NEW BEST: epoch 3 valid F1 85.06 best test F1 81.55
NEW BEST: epoch 4 valid F1 86.97 best test F1 83.42
NEW BEST: epoch 5 valid F1 89.2 best test F1 85.05
NEW BEST: epoch 6 valid F1 89.86 best test F1 86.12
NEW BEST: epoch 7 valid F1 91.53 best test F1 88.72
NEW BEST: epoch 8 valid F1 92.68 best test F1 89.1
NEW BEST: epoch 9 valid F1 93.35 best test F1 89.81
NEW BEST: epoch 10 valid F1 94.17 best test F1 91.22
NEW BEST: epoch 11 valid F1 94.61 best test F1 91.53
NEW BEST: epoch 12 valid F1 94.64 best test F1 92.02
NEW BEST: epoch 13 valid F1 95.1 best test F1 92.53
NEW BEST: epoch 14 valid F1 95.2 best test F1 93.07

In [3]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.42 best test F1 64.7
NEW BEST: epoch 1 valid F1 80.53 best test F1 77.05
NEW BEST: epoch 2 valid F1 84.25 best test F1 81.63
NEW BEST: epoch 3 valid F1 87.16 best test F1 83.29
NEW BEST: epoch 4 valid F1 88.94 best test F1 85.25
NEW BEST: epoch 5 valid F1 90.91 best test F1 86.3
NEW BEST: epoch 6 valid F1 91.6 best test F1 87.7
NEW BEST: epoch 7 valid F1 92.58 best test F1 89.88
NEW BEST: epoch 8 valid F1 93.29 best test F1 90.82
NEW BEST: epoch 9 valid F1 93.92 best test F1 91.35
NEW BEST: epoch 10 valid F1 94.35 best test F1 92.43
NEW BEST: epoch 11 valid F1 94.74 best test F1 92.54
NEW BEST: epoch 12 valid F1 94.75 best test F1 93.03
NEW BEST: epoch 13 valid F1 95.12 best test F1 92.6
NEW BEST: epoch 14 valid F1 95.18 best test F1 93

In [2]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 64.94 best test F1 61.36
NEW BEST: epoch 1 valid F1 75.74 best test F1 71.72
NEW BEST: epoch 2 valid F1 80.7 best test F1 77.78
NEW BEST: epoch 3 valid F1 84.52 best test F1 81.29
NEW BEST: epoch 4 valid F1 87.61 best test F1 83.27
NEW BEST: epoch 5 valid F1 89.0 best test F1 85.01
NEW BEST: epoch 6 valid F1 90.37 best test F1 85.73
NEW BEST: epoch 7 valid F1 91.26 best test F1 88.45
NEW BEST: epoch 8 valid F1 92.4 best test F1 88.9
NEW BEST: epoch 9 valid F1 93.37 best test F1 89.92
NEW BEST: epoch 10 valid F1 93.79 best test F1 90.83
NEW BEST: epoch 11 valid F1 94.01 best test F1 91.46
NEW BEST: epoch 12 valid F1 94.98 best test F1 91.85
NEW BEST: epoch 13 valid F1 95.39 best test F1 91.65

NEW BEST: epoch 15 valid F1 95.48 best test F1 9

In [2]:
emb, idx2label, idx2word = test_rnnslu(nhidden=100, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=False)

verbose: True
normal: False
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.33 best test F1 64.43
NEW BEST: epoch 1 valid F1 81.03 best test F1 76.62
NEW BEST: epoch 2 valid F1 84.58 best test F1 80.76
NEW BEST: epoch 3 valid F1 86.66 best test F1 82.81
NEW BEST: epoch 4 valid F1 89.12 best test F1 84.93
NEW BEST: epoch 5 valid F1 90.5 best test F1 86.67
NEW BEST: epoch 6 valid F1 90.84 best test F1 87.63
NEW BEST: epoch 7 valid F1 92.2 best test F1 89.35
NEW BEST: epoch 8 valid F1 93.06 best test F1 90.11
NEW BEST: epoch 9 valid F1 93.59 best test F1 91.04
NEW BEST: epoch 10 valid F1 94.24 best test F1 91.33
NEW BEST: epoch 11 valid F1 94.5 best test F1 92.14
NEW BEST: epoch 12 valid F1 94.84 best test F1 92.3
NEW BEST: epoch 13 valid F1 95.14 best test F1 92.83
NEW BEST: epoch 14 valid F1 95.4 best test F1 92

#### Bullet 3: Implement RNN with two hidden layers

My test F1 results range from 90.61% to 93.54%. I used the same parameter combinations in bullets 1 and 2 plus the extra parameters for nhidden1 for which I tried all combinations of 100 and 200 hidden units. The best model also had the model hidden units in both layers: nhidden0=200, nhidden1=200, win=5, emb_dimension=100. Meanwhile the worst model had the least number of hidden units between the two layers of all the models I tested: nhidden0=100, nhidden1=100, win=5, emb_dimension=50, which leads me to believe that generally more hidden units leads to better test performance.

In [1]:
test_rnnslu2(nhidden0=200, nhidden1=100, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

Using gpu device 0: GRID K520 (CNMeM is disabled, CuDNN 3007)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
nhidden0: 100
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.29 best test F1 50.79
NEW BEST: epoch 1 valid F1 67.55 best test F1 65.26
NEW BEST: epoch 2 valid F1 72.16 best test F1 69.47
NEW BEST: epoch 3 valid F1 80.43 best test F1 76.01
NEW BEST: epoch 4 valid F1 83.84 best test F1 80.23
NEW BEST: epoch 5 valid F1 86.04 best test F1 83.3
NEW BEST: epoch 6 valid F1 87.82 best test F1 84.36
NEW BEST: epoch 7 valid F1 89.48 best test F1 86.31
NEW BEST: epoch 8 valid F1 89.79 best test F1 87.9

NEW BEST: epoch 10 valid F1 91.17 best test F1 88.46
NEW BEST: epoch 11 valid F1 91.91 best test F1 89.19
NEW BEST: epoch 12 valid F1 92.64 best test F1 89.8
NEW BEST: epoch 13 valid F1 93.18 best test F1 89.92


NEW BEST: epoch 16 valid F1 93.32 best test F1 90.07
NEW BEST: epoch 17 valid F1 94

In [4]:
test_rnnslu2(nhidden0=200, nhidden1=100, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 200
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 57.41 best test F1 53.38
NEW BEST: epoch 1 valid F1 69.34 best test F1 68.07
NEW BEST: epoch 2 valid F1 78.22 best test F1 74.84
NEW BEST: epoch 3 valid F1 82.43 best test F1 78.06
NEW BEST: epoch 4 valid F1 85.0 best test F1 81.08
NEW BEST: epoch 5 valid F1 86.81 best test F1 83.32
NEW BEST: epoch 6 valid F1 88.2 best test F1 84.77
NEW BEST: epoch 7 valid F1 89.22 best test F1 86.15
NEW BEST: epoch 8 valid F1 90.15 best test F1 87.73
NEW BEST: epoch 9 valid F1 90.79 best test F1 87.72
NEW BEST: epoch 10 valid F1 91.28 best test F1 88.91
NEW BEST: epoch 11 valid F1 92.39 best test F1 90.35
NEW BEST: epoch 12 valid F1 93.07 best test F1 90.9
NEW BEST: epoch 13 valid F1 93.99 best test F1 90.57

NEW BEST: epoch 15 valid F1 94.6

In [6]:
test_rnnslu2(nhidden0=200, nhidden1=100, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 200
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.14 best test F1 54.02
NEW BEST: epoch 1 valid F1 69.46 best test F1 66.44
NEW BEST: epoch 2 valid F1 75.6 best test F1 73.11
NEW BEST: epoch 3 valid F1 80.04 best test F1 77.48
NEW BEST: epoch 4 valid F1 84.54 best test F1 79.97
NEW BEST: epoch 5 valid F1 85.88 best test F1 81.39
NEW BEST: epoch 6 valid F1 86.47 best test F1 84.33
NEW BEST: epoch 7 valid F1 88.64 best test F1 85.75
NEW BEST: epoch 8 valid F1 89.86 best test F1 86.83
NEW BEST: epoch 9 valid F1 90.58 best test F1 86.95
NEW BEST: epoch 10 valid F1 90.74 best test F1 87.76
NEW BEST: epoch 11 valid F1 91.58 best test F1 88.9
NEW BEST: epoch 12 valid F1 92.58 best test F1 89.54
NEW BEST: epoch 13 valid F1 92.63 best test F1 89.27

NEW BEST: epoch 15 valid F1 93.6

In [2]:
test_rnnslu2(nhidden0=200, nhidden1=100, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 200
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.88 best test F1 54.86
NEW BEST: epoch 1 valid F1 69.13 best test F1 66.58
NEW BEST: epoch 2 valid F1 76.83 best test F1 74.62
NEW BEST: epoch 3 valid F1 81.64 best test F1 77.96
NEW BEST: epoch 4 valid F1 85.24 best test F1 80.87
NEW BEST: epoch 5 valid F1 87.16 best test F1 83.92
NEW BEST: epoch 6 valid F1 87.88 best test F1 84.82
NEW BEST: epoch 7 valid F1 89.33 best test F1 86.58
NEW BEST: epoch 8 valid F1 90.35 best test F1 87.7
NEW BEST: epoch 9 valid F1 90.47 best test F1 87.55
NEW BEST: epoch 10 valid F1 91.57 best test F1 88.59
NEW BEST: epoch 11 valid F1 92.26 best test F1 89.87
NEW BEST: epoch 12 valid F1 92.62 best test F1 89.46
NEW BEST: epoch 13 valid F1 92.77 best test F1 89.62

NEW BEST: epoch 15 valid F1 93

In [3]:
test_rnnslu2(nhidden0=200, nhidden1=200, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 200
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 58.44 best test F1 54.58
NEW BEST: epoch 1 valid F1 71.89 best test F1 68.75
NEW BEST: epoch 2 valid F1 78.18 best test F1 75.83
NEW BEST: epoch 3 valid F1 82.38 best test F1 78.77
NEW BEST: epoch 4 valid F1 84.96 best test F1 81.43
NEW BEST: epoch 5 valid F1 87.53 best test F1 84.0
NEW BEST: epoch 6 valid F1 87.78 best test F1 84.71
NEW BEST: epoch 7 valid F1 89.93 best test F1 87.38
NEW BEST: epoch 8 valid F1 90.65 best test F1 88.12
NEW BEST: epoch 9 valid F1 91.26 best test F1 88.77
NEW BEST: epoch 10 valid F1 92.36 best test F1 89.68
NEW BEST: epoch 11 valid F1 93.14 best test F1 90.45
NEW BEST: epoch 12 valid F1 93.4 best test F1 91.08
NEW BEST: epoch 13 valid F1 93.81 best test F1 91.74

NEW BEST: epoch 15 valid F1 94.2

In [4]:
test_rnnslu2(nhidden0=200, nhidden1=200, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 200
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 60.0 best test F1 57.72
NEW BEST: epoch 1 valid F1 73.89 best test F1 70.91
NEW BEST: epoch 2 valid F1 80.13 best test F1 77.09
NEW BEST: epoch 3 valid F1 83.26 best test F1 79.86
NEW BEST: epoch 4 valid F1 86.0 best test F1 83.18
NEW BEST: epoch 5 valid F1 88.35 best test F1 84.93
NEW BEST: epoch 6 valid F1 89.48 best test F1 85.83
NEW BEST: epoch 7 valid F1 90.54 best test F1 87.82
NEW BEST: epoch 8 valid F1 91.44 best test F1 89.54
NEW BEST: epoch 9 valid F1 93.11 best test F1 90.79
NEW BEST: epoch 10 valid F1 93.13 best test F1 90.4
NEW BEST: epoch 11 valid F1 94.09 best test F1 91.27

NEW BEST: epoch 13 valid F1 94.52 best test F1 91.57
NEW BEST: epoch 14 valid F1 94.59 best test F1 90.7
NEW BEST: epoch 15 valid F1 94.78

In [5]:
test_rnnslu2(nhidden0=200, nhidden1=200, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 200
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 59.13 best test F1 55.26
NEW BEST: epoch 1 valid F1 73.02 best test F1 69.98
NEW BEST: epoch 2 valid F1 79.05 best test F1 75.2
NEW BEST: epoch 3 valid F1 83.2 best test F1 79.71
NEW BEST: epoch 4 valid F1 85.33 best test F1 82.18
NEW BEST: epoch 5 valid F1 88.37 best test F1 84.69
NEW BEST: epoch 6 valid F1 89.38 best test F1 86.15
NEW BEST: epoch 7 valid F1 90.53 best test F1 87.32
NEW BEST: epoch 8 valid F1 91.17 best test F1 88.55

NEW BEST: epoch 10 valid F1 92.71 best test F1 90.06

NEW BEST: epoch 12 valid F1 93.5 best test F1 90.61
NEW BEST: epoch 13 valid F1 94.38 best test F1 91.39

NEW BEST: epoch 15 valid F1 94.65 best test F1 90.72

NEW BEST: epoch 17 valid F1 95.06 best test F1 91.94
NEW BEST: epoch 18 valid F1 9

In [6]:
test_rnnslu2(nhidden0=200, nhidden1=200, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 200
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 61.68 best test F1 58.47
NEW BEST: epoch 1 valid F1 72.49 best test F1 68.89
NEW BEST: epoch 2 valid F1 79.42 best test F1 75.76
NEW BEST: epoch 3 valid F1 83.23 best test F1 79.43
NEW BEST: epoch 4 valid F1 85.5 best test F1 81.23
NEW BEST: epoch 5 valid F1 87.65 best test F1 84.0
NEW BEST: epoch 6 valid F1 88.77 best test F1 85.1
NEW BEST: epoch 7 valid F1 90.33 best test F1 88.23
NEW BEST: epoch 8 valid F1 91.25 best test F1 88.94
NEW BEST: epoch 9 valid F1 91.56 best test F1 89.02
NEW BEST: epoch 10 valid F1 92.74 best test F1 90.15
NEW BEST: epoch 11 valid F1 93.39 best test F1 91.41
NEW BEST: epoch 12 valid F1 93.49 best test F1 91.53
NEW BEST: epoch 13 valid F1 94.26 best test F1 91.4

NEW BEST: epoch 15 valid F1 94.91

In [7]:
test_rnnslu2(nhidden0=100, nhidden1=100, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 100
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 53.4 best test F1 49.51
NEW BEST: epoch 1 valid F1 64.92 best test F1 61.78
NEW BEST: epoch 2 valid F1 70.61 best test F1 68.24
NEW BEST: epoch 3 valid F1 76.65 best test F1 74.88
NEW BEST: epoch 4 valid F1 81.57 best test F1 77.87
NEW BEST: epoch 5 valid F1 83.49 best test F1 79.59
NEW BEST: epoch 6 valid F1 85.47 best test F1 82.14
NEW BEST: epoch 7 valid F1 86.81 best test F1 83.41
NEW BEST: epoch 8 valid F1 87.67 best test F1 84.95
NEW BEST: epoch 9 valid F1 88.25 best test F1 85.46
NEW BEST: epoch 10 valid F1 89.43 best test F1 86.99
NEW BEST: epoch 11 valid F1 90.88 best test F1 87.61
NEW BEST: epoch 12 valid F1 91.69 best test F1 87.92


NEW BEST: epoch 15 valid F1 92.55 best test F1 88.42

NEW BEST: epoch 17 valid F1 9

In [8]:
test_rnnslu2(nhidden0=100, nhidden1=100, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 100
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.01 best test F1 52.39
NEW BEST: epoch 1 valid F1 62.86 best test F1 60.66
NEW BEST: epoch 2 valid F1 73.33 best test F1 70.42
NEW BEST: epoch 3 valid F1 78.5 best test F1 74.9
NEW BEST: epoch 4 valid F1 82.64 best test F1 79.23
NEW BEST: epoch 5 valid F1 84.99 best test F1 81.25
NEW BEST: epoch 6 valid F1 86.78 best test F1 83.68
NEW BEST: epoch 7 valid F1 88.36 best test F1 84.13
NEW BEST: epoch 8 valid F1 88.66 best test F1 86.61
NEW BEST: epoch 9 valid F1 89.46 best test F1 86.92
NEW BEST: epoch 10 valid F1 90.65 best test F1 88.04
NEW BEST: epoch 11 valid F1 91.32 best test F1 88.85
NEW BEST: epoch 12 valid F1 92.26 best test F1 89.0


NEW BEST: epoch 15 valid F1 92.63 best test F1 89.26
NEW BEST: epoch 16 valid F1 92.

In [9]:
test_rnnslu2(nhidden0=100, nhidden1=100, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 100
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.36 best test F1 51.1
NEW BEST: epoch 1 valid F1 63.41 best test F1 61.97
NEW BEST: epoch 2 valid F1 71.64 best test F1 69.63
NEW BEST: epoch 3 valid F1 77.88 best test F1 75.23
NEW BEST: epoch 4 valid F1 81.73 best test F1 78.14
NEW BEST: epoch 5 valid F1 83.66 best test F1 80.01
NEW BEST: epoch 6 valid F1 85.36 best test F1 82.12
NEW BEST: epoch 7 valid F1 86.68 best test F1 82.6
NEW BEST: epoch 8 valid F1 86.86 best test F1 83.99

NEW BEST: epoch 10 valid F1 88.81 best test F1 86.02
NEW BEST: epoch 11 valid F1 89.85 best test F1 86.99
NEW BEST: epoch 12 valid F1 90.93 best test F1 87.68
NEW BEST: epoch 13 valid F1 91.27 best test F1 88.17

NEW BEST: epoch 15 valid F1 91.41 best test F1 88.8
NEW BEST: epoch 16 valid F1 91.

In [10]:
test_rnnslu2(nhidden0=100, nhidden1=100, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 100
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.71 best test F1 53.51
NEW BEST: epoch 1 valid F1 63.2 best test F1 61.32
NEW BEST: epoch 2 valid F1 71.38 best test F1 69.58
NEW BEST: epoch 3 valid F1 78.11 best test F1 75.37
NEW BEST: epoch 4 valid F1 81.06 best test F1 76.86
NEW BEST: epoch 5 valid F1 83.41 best test F1 80.21
NEW BEST: epoch 6 valid F1 84.89 best test F1 82.94
NEW BEST: epoch 7 valid F1 86.85 best test F1 83.81
NEW BEST: epoch 8 valid F1 87.21 best test F1 85.38
NEW BEST: epoch 9 valid F1 88.1 best test F1 85.2
NEW BEST: epoch 10 valid F1 89.78 best test F1 87.29
NEW BEST: epoch 11 valid F1 90.21 best test F1 88.47
NEW BEST: epoch 12 valid F1 90.76 best test F1 88.79

NEW BEST: epoch 14 valid F1 90.91 best test F1 87.61
NEW BEST: epoch 15 valid F1 92.3

In [11]:
test_rnnslu2(nhidden0=100, nhidden1=200, win=7, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 100
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.29 best test F1 50.79
NEW BEST: epoch 1 valid F1 67.55 best test F1 65.26
NEW BEST: epoch 2 valid F1 72.16 best test F1 69.47
NEW BEST: epoch 3 valid F1 80.43 best test F1 76.01
NEW BEST: epoch 4 valid F1 83.84 best test F1 80.23
NEW BEST: epoch 5 valid F1 86.04 best test F1 83.3
NEW BEST: epoch 6 valid F1 87.82 best test F1 84.36
NEW BEST: epoch 7 valid F1 89.48 best test F1 86.31
NEW BEST: epoch 8 valid F1 89.79 best test F1 87.9

NEW BEST: epoch 10 valid F1 91.17 best test F1 88.46
NEW BEST: epoch 11 valid F1 91.91 best test F1 89.19
NEW BEST: epoch 12 valid F1 92.64 best test F1 89.8
NEW BEST: epoch 13 valid F1 93.18 best test F1 89.92


NEW BEST: epoch 16 valid F1 93.32 best test F1 90.07
NEW BEST: epoch 17 valid F1 94

In [12]:
test_rnnslu2(nhidden0=100, nhidden1=200, win=7, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 7
nhidden0: 100
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.25 best test F1 52.24
NEW BEST: epoch 1 valid F1 67.25 best test F1 64.49
NEW BEST: epoch 2 valid F1 75.67 best test F1 72.72
NEW BEST: epoch 3 valid F1 80.21 best test F1 75.73
NEW BEST: epoch 4 valid F1 83.99 best test F1 80.16
NEW BEST: epoch 5 valid F1 86.53 best test F1 82.85
NEW BEST: epoch 6 valid F1 87.84 best test F1 84.41
NEW BEST: epoch 7 valid F1 88.85 best test F1 86.48
NEW BEST: epoch 8 valid F1 90.02 best test F1 88.46
NEW BEST: epoch 9 valid F1 90.4 best test F1 86.85
NEW BEST: epoch 10 valid F1 92.03 best test F1 89.68
NEW BEST: epoch 11 valid F1 92.37 best test F1 89.83
NEW BEST: epoch 12 valid F1 92.55 best test F1 89.74
NEW BEST: epoch 13 valid F1 93.54 best test F1 90.54

NEW BEST: epoch 15 valid F1 93

In [13]:
test_rnnslu2(nhidden0=100, nhidden1=200, win=5, emb_dimension=50, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 100
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.26 best test F1 51.71
NEW BEST: epoch 1 valid F1 66.07 best test F1 64.66
NEW BEST: epoch 2 valid F1 74.1 best test F1 71.73
NEW BEST: epoch 3 valid F1 78.11 best test F1 75.13
NEW BEST: epoch 4 valid F1 82.76 best test F1 79.49
NEW BEST: epoch 5 valid F1 85.53 best test F1 82.67
NEW BEST: epoch 6 valid F1 87.32 best test F1 82.92
NEW BEST: epoch 7 valid F1 89.09 best test F1 85.55
NEW BEST: epoch 8 valid F1 89.45 best test F1 87.58

NEW BEST: epoch 10 valid F1 91.16 best test F1 88.74
NEW BEST: epoch 11 valid F1 91.95 best test F1 89.03
NEW BEST: epoch 12 valid F1 93.1 best test F1 90.19
NEW BEST: epoch 13 valid F1 93.41 best test F1 90.5

NEW BEST: epoch 15 valid F1 93.68 best test F1 90.72
NEW BEST: epoch 16 valid F1 93.

In [14]:
test_rnnslu2(nhidden0=100, nhidden1=200, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
nhidden0: 100
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 55.66 best test F1 52.18
NEW BEST: epoch 1 valid F1 67.55 best test F1 65.39
NEW BEST: epoch 2 valid F1 76.78 best test F1 73.59
NEW BEST: epoch 3 valid F1 80.18 best test F1 76.74
NEW BEST: epoch 4 valid F1 83.84 best test F1 81.68
NEW BEST: epoch 5 valid F1 86.02 best test F1 83.2
NEW BEST: epoch 6 valid F1 87.57 best test F1 84.39
NEW BEST: epoch 7 valid F1 89.46 best test F1 86.4
NEW BEST: epoch 8 valid F1 89.74 best test F1 87.52

NEW BEST: epoch 10 valid F1 90.92 best test F1 88.51
NEW BEST: epoch 11 valid F1 91.96 best test F1 89.53
NEW BEST: epoch 12 valid F1 92.79 best test F1 90.53
NEW BEST: epoch 13 valid F1 93.77 best test F1 89.61



NEW BEST: epoch 17 valid F1 94.24 best test F1 91.5
NEW BEST: epoch 18 valid F1 

#### Bullet 4: Experiment with word embeddings

Here I look at differences in word embeddings that are similar, as well as embeddings of specific words that are similar.

##### Similarities between differences of word embeddings

In [3]:
emb, idx2label, idx2word= test_rnnslu(nhidden=200, win=5, emb_dimension=100, verbose=True, nepochs=25, normal=True)

verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 25
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.76 best test F1 68.63
NEW BEST: epoch 1 valid F1 80.95 best test F1 78.45
NEW BEST: epoch 2 valid F1 84.54 best test F1 81.6
NEW BEST: epoch 3 valid F1 87.75 best test F1 84.64
NEW BEST: epoch 4 valid F1 89.6 best test F1 86.41
NEW BEST: epoch 5 valid F1 91.2 best test F1 87.78

NEW BEST: epoch 7 valid F1 92.58 best test F1 90.04
NEW BEST: epoch 8 valid F1 93.78 best test F1 91.27
NEW BEST: epoch 9 valid F1 94.37 best test F1 91.22
NEW BEST: epoch 10 valid F1 95.04 best test F1 92.06
NEW BEST: epoch 11 valid F1 95.37 best test F1 92.75
NEW BEST: epoch 12 valid F1 95.58 best test F1 92.72


NEW BEST: epoch 15 valid F1 96.35 best test F1 93.51

NEW BEST: epoch 17 valid F1 96.63 best test F1 93.45



NEW BEST: epoch 21 valid F1 96.94 best t

In [11]:
from scipy import spatial

def diff_matrix(r1,r2,r3,r4):
    diff={}
    for i in range(r1,r2):
        for j in range(r3,r4):
            if i<j:
                diff[(i,j)]=numpy.abs(numpy.asarray(emb[i])-numpy.asarray(emb[j]))
        
    diff_mat=numpy.zeros((len(diff), len(diff)))+numpy.inf
    for n, i in enumerate(diff.keys()):
        for m, j in enumerate(diff.keys()[n+1:]):
            diff_mat[n, n+m+1]=spatial.distance.euclidean(diff[i], diff[j])
        
    for n, i in enumerate(numpy.argmin(diff_mat,0)):
        if diff_mat[i,n]<0.75:
            if ('DIGIT' not in idx2word[diff.keys()[n][0]]) and ('DIGIT' not in idx2word[diff.keys()[i][0]]) and ('DIGIT' not in idx2word[diff.keys()[i][1]]) and ('UNK' not in idx2word[diff.keys()[i][0]]) and ('UNK' not in idx2word[diff.keys()[i][1]]):
                if (diff.keys()[n][0]!=diff.keys()[i][0]) and (diff.keys()[n][1]!=diff.keys()[i][1]) and (diff.keys()[n][1]!=diff.keys()[i][0]) and (diff.keys()[n][0]!=diff.keys()[i][1]):
                    print '\"%s\" minus \"%s\" approx. equals \"%s\" minus \"%s\" with distance %s' % (idx2word[diff.keys()[n][0]], idx2word[diff.keys()[n][1]], idx2word[diff.keys()[i][0]], idx2word[diff.keys()[i][1]], diff_mat[i,n]) 


In [12]:
diff_matrix(10,50,30,70)

"after" minus "approximately" approx. equals "ap" minus "ap80" with distance 0.736545324326
"ap" minus "arriving" approx. equals "ap57" minus "arrive" with distance 0.650590896606
"airports" minus "any" approx. equals "ap" minus "ap80" with distance 0.748420715332
"april" minus "august" approx. equals "ap" minus "ap57" with distance 0.611584961414
"america" minus "american" approx. equals "april" minus "august" with distance 0.651915252209
"again" minus "at" approx. equals "america" minus "american" with distance 0.736275851727
"an" minus "at" approx. equals "america" minus "american" with distance 0.653774142265
"and" minus "any" approx. equals "april" minus "august" with distance 0.737818241119
"after" minus "around" approx. equals "april" minus "august" with distance 0.668402373791
"and" minus "anywhere" approx. equals "after" minus "around" with distance 0.746643483639
"about" minus "around" approx. equals "ap" minus "ap57" with distance 0.60206925869
"angeles" minus "area" approx.

In [13]:
diff_matrix(200,250,300,350)

"indianapolis" minus "newark" approx. equals "july" minus "november" with distance 0.739346265793
"houston" minus "minneapolis" approx. equals "july" minus "november" with distance 0.720242977142
"friday" minus "monday" approx. equals "july" minus "november" with distance 0.583707809448
"houston" minus "milwaukee" approx. equals "july" minus "november" with distance 0.678772866726
"friday" minus "november" approx. equals "june" minus "monday" with distance 0.674577534199
"indianapolis" minus "nashville" approx. equals "january" minus "november" with distance 0.72392731905
"houston" minus "newark" approx. equals "january" minus "november" with distance 0.685251295567
"international" minus "mitchell" approx. equals "january" minus "november" with distance 0.603928625584
"friday" minus "minneapolis" approx. equals "houston" minus "monday" with distance 0.732950866222
"guardia" minus "mitchell" approx. equals "january" minus "november" with distance 0.579018652439
"june" minus "november" a

In [14]:
diff_matrix(100,150,200,250)

"columbus" minus "indianapolis" approx. equals "denver" minus "houston" with distance 0.735874593258
"co" minus "hp" approx. equals "denver" minus "houston" with distance 0.664528608322
"city" minus "fort" approx. equals "denver" minus "houston" with distance 0.745087683201
"chicago" minus "indianapolis" approx. equals "denver" minus "houston" with distance 0.634911715984
"daily" minus "fort" approx. equals "denver" minus "houston" with distance 0.740564346313
"december" minus "january" approx. equals "denver" minus "houston" with distance 0.650330424309
"denver" minus "indianapolis" approx. equals "december" minus "january" with distance 0.693463921547
"december" minus "hp" approx. equals "denver" minus "houston" with distance 0.745889723301
"codes" minus "how" approx. equals "denver" minus "houston" with distance 0.655982017517
"detroit" minus "indianapolis" approx. equals "denver" minus "houston" with distance 0.736925840378
"depart" minus "from" approx. equals "december" minus "jan

In [16]:
diff_matrix(75,125,225,275)

"cheapest" minus "houston" approx. equals "chicago" minus "less" with distance 0.738542973995
"bwi" minus "jfk" approx. equals "columbus" minus "indianapolis" with distance 0.742628753185
"charlotte" minus "indianapolis" approx. equals "bwi" minus "jfk" with distance 0.704390823841
"cheapest" minus "less" approx. equals "charlotte" minus "indianapolis" with distance 0.676905691624
"cheap" minus "last" approx. equals "co" minus "hp" with distance 0.717147767544
"city" minus "lake" approx. equals "co" minus "hp" with distance 0.687451839447


##### Similarities between word embeddings

In [ ]:
from scipy import spatial
nwords=emb.shape[0]
dist=numpy.zeros((nwords, nwords))
for i in range(nwords):
    for j in range(nwords):
        if i!=j:
            dist[i,j]=spatial.distance.euclidean(emb[i], emb[j])
        elif i==j:
            dist[i,j]=numpy.inf

In [21]:
print 'Words that are withing 0.5 euclidean distance of each other:'
for n, i in enumerate(numpy.argmin(dist,0)):
    if dist[n,i]<0.5:
        print idx2word[n],'is close to', idx2word[i],'with distance', dist[n,i]

Words that are withing 0.5 euclidean distance of each other:
american is close to delta with distance 0.4157089293
arrive is close to arriving with distance 0.371802091599
arrives is close to arrive with distance 0.495740056038
arriving is close to arrive with distance 0.371802091599
august is close to june with distance 0.458989500999
breakfast is close to dinner with distance 0.332106739283
car is close to rental with distance 0.284863650799
continental is close to american with distance 0.422604560852
december is close to august with distance 0.475253432989
delta is close to american with distance 0.4157089293
dinner is close to breakfast with distance 0.332106739283
dl is close to twa with distance 0.477683752775
eastern is close to united with distance 0.496108680964
friday is close to sunday with distance 0.483597040176
june is close to august with distance 0.458989500999
limousine is close to car with distance 0.456346899271
meal is close to meals with distance 0.425383925438
me

### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

I was able to achieve 100% test accuracy for 8-bits, but only 99% for 12-bit. The tests are really sensitive to initalization. And it seems the more bits the more difficult it is to get the algorithm to 100% learn the parity function.

In [ ]:
test_mlp_parity(learning_rate=0.1, L1_reg=0.00, L2_reg=0.000, n_epochs=4000,
             batch_size=20, n_hidden=17, n_hiddenLayers=1,
             verbose=True, nbit=8, activation=T.nnet.sigmoid, nsamples=[1000,500,100])

... building the model
... training
training @ iter =  0
epoch 1, minibatch 50/50, validation error 52.800000 %
     epoch 1, minibatch 50/50, test error of best model 54.000000 %
epoch 2, minibatch 50/50, validation error 51.400000 %
     epoch 2, minibatch 50/50, test error of best model 50.000000 %
training @ iter =  100
epoch 3, minibatch 50/50, validation error 51.000000 %
     epoch 3, minibatch 50/50, test error of best model 51.000000 %
epoch 4, minibatch 50/50, validation error 51.000000 %
training @ iter =  200
epoch 5, minibatch 50/50, validation error 51.000000 %
epoch 6, minibatch 50/50, validation error 51.000000 %
training @ iter =  300
epoch 7, minibatch 50/50, validation error 50.600000 %
     epoch 7, minibatch 50/50, test error of best model 50.000000 %
epoch 8, minibatch 50/50, validation error 50.600000 %
training @ iter =  400
epoch 9, minibatch 50/50, validation error 50.600000 %
epoch 10, minibatch 50/50, validation error 50.600000 %
training @ iter =  500
epoch

The training process for function test_mlp_parity ran for 1.56m


In [6]:
test_mlp_parity(learning_rate=0.2, L1_reg=0.0, L2_reg=0.00001, n_epochs=20000,
             batch_size=20, n_hidden=75, n_hiddenLayers=1,
             verbose=False, nbit=12, activation=T.tanh, nsamples=[1000,500,100])

... building the model
... training
Optimization complete.
Best validation score of 1.200000 % obtained at iteration 545150, with test performance 1.000000 %


The training process for function test_mlp_parity ran for 11.61m


#### Bullet 2: Deep MLP

I was able to learn parity for both the 8 and 12-bit models for deep MLP.

In [65]:
test_mlp_parity(learning_rate=0.02, L1_reg=0.0, L2_reg=0.000, n_epochs=4000,
             batch_size=20, n_hidden=5, n_hiddenLayers=9,
             verbose=True, nbit=8, activation=T.tanh, nsamples=[1000,500,100])

... building the model
... training
training @ iter =  0
epoch 1, minibatch 50/50, validation error 47.600000 %
     epoch 1, minibatch 50/50, test error of best model 44.000000 %
epoch 2, minibatch 50/50, validation error 49.200000 %
training @ iter =  100
epoch 3, minibatch 50/50, validation error 49.800000 %
epoch 4, minibatch 50/50, validation error 51.800000 %
training @ iter =  200
epoch 5, minibatch 50/50, validation error 52.400000 %
epoch 6, minibatch 50/50, validation error 52.400000 %
training @ iter =  300
epoch 7, minibatch 50/50, validation error 52.200000 %
epoch 8, minibatch 50/50, validation error 52.000000 %
training @ iter =  400
epoch 9, minibatch 50/50, validation error 52.600000 %
epoch 10, minibatch 50/50, validation error 51.600000 %
training @ iter =  500
epoch 11, minibatch 50/50, validation error 51.600000 %
epoch 12, minibatch 50/50, validation error 51.600000 %
training @ iter =  600
epoch 13, minibatch 50/50, validation error 51.800000 %
epoch 14, minibatc

The training process for function test_mlp_parity ran for 1.29m


In [86]:
test_mlp_parity(learning_rate=0.02, L1_reg=0.0, L2_reg=0.000, n_epochs=4000,
             batch_size=20, n_hidden=7, n_hiddenLayers=10,
             verbose=True, nbit=12, activation=T.tanh, nsamples=[1000,500,100])

... building the model
... training
training @ iter =  0
epoch 1, minibatch 50/50, validation error 50.600000 %
     epoch 1, minibatch 50/50, test error of best model 51.000000 %
epoch 2, minibatch 50/50, validation error 50.600000 %
training @ iter =  100
epoch 3, minibatch 50/50, validation error 50.600000 %
epoch 4, minibatch 50/50, validation error 50.600000 %
training @ iter =  200
epoch 5, minibatch 50/50, validation error 50.600000 %
epoch 6, minibatch 50/50, validation error 50.600000 %
training @ iter =  300
epoch 7, minibatch 50/50, validation error 50.600000 %
epoch 8, minibatch 50/50, validation error 50.600000 %
training @ iter =  400
epoch 9, minibatch 50/50, validation error 50.600000 %
epoch 10, minibatch 50/50, validation error 50.600000 %
training @ iter =  500
epoch 11, minibatch 50/50, validation error 50.600000 %
epoch 12, minibatch 50/50, validation error 50.600000 %
training @ iter =  600
epoch 13, minibatch 50/50, validation error 50.600000 %
epoch 14, minibatc

The training process for function test_mlp_parity ran for 1.46m


#### Bullet 3: RNN

I tested an 8-bit and 12-bit model and was able to get 100% test F1. 

In [2]:
test_rnn_parity(nbits=8, nsamples=[1000, 500, 100], lr=0.2, nhidden=3, win=7, nepochs=300)

verbose: True
win: 7
savemodel: False
nsamples: [1000, 500, 100]
nhidden: 3
nbits: 8
decay: True
lr: 0.2
folder: ../result
nepochs: 300
... loading the dataset
... building the model
... training
(67, 506)
('NEW BEST: epoch', 0, 'training F1', 0.53606411398040965, 'valid F1', 0.56437389770723112, 'best test F1', 0.5)
(57, 506)

(57, 506)

(57, 506)

(80, 506)
('NEW BEST: epoch', 4, 'training F1', 0.62090981644054266, 'valid F1', 0.597444089456869, 'best test F1', 0.57600000000000007)
(77, 506)
('NEW BEST: epoch', 5, 'training F1', 0.58231954582319545, 'valid F1', 0.62745098039215697, 'best test F1', 0.57377049180327877)
(50, 506)

(27, 506)

(11, 506)

(11, 506)

(11, 506)

(23, 506)

(26, 506)

(45, 506)
('NEW BEST: epoch', 13, 'training F1', 1.0, 'valid F1', 1.0, 'best test F1', 1.0)
(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 506)

(45, 

In [10]:
test_rnn_parity(nbits=12, nsamples=[1000, 500, 100], lr=0.3, nhidden=6, win=1, nepochs=300)

verbose: True
win: 1
savemodel: False
nsamples: [1000, 500, 100]
nhidden: 6
nbits: 12
decay: True
lr: 0.3
folder: ../result
nepochs: 300
... loading the dataset
... building the model
... training
(48, 521)
('NEW BEST: epoch', 0, 'training F1', 0.4837117472852912, 'valid F1', 0.5357142857142857, 'best test F1', 0.47422680412371138)
(48, 521)

(48, 521)

(48, 521)

(100, 521)
('NEW BEST: epoch', 4, 'training F1', 0.68507560815253132, 'valid F1', 0.66131191432396252, 'best test F1', 0.65771812080536907)
(100, 521)

(100, 521)

(100, 521)

(100, 521)

(100, 521)

(100, 521)

(100, 521)

(100, 521)

(77, 521)

(47, 521)
('NEW BEST: epoch', 14, 'training F1', 0.99132111861137884, 'valid F1', 0.99389002036659879, 'best test F1', 0.97916666666666663)
(49, 521)
('NEW BEST: epoch', 15, 'training F1', 1.0, 'valid F1', 1.0, 'best test F1', 1.0)
(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

(49, 521)

